pkg/trace/writer/newwriter/sender.go

```diff
@@ -0,0 +1,458 @@
+	}
+}
+
+// drainQueue drains the entire queue and returns all the payloads that were in it.
+func (q *sender) drainQueue() []*payload {
+	q.mu.Lock()
+	defer q.mu.Unlock()
+	var payloads []*payload
+	for q.list.Len() > 0 {
+		v := q.list.Remove(q.list.Front())
+		payloads = append(payloads, v.(*payload))
+	}
+	q.size = 0
+	return payloads
+}
+
+// sendPayloads concurrently sends the given list of payloads. It returns
+// the number of payloads that were added back onto the queue to be retried
+// again later due to an error.
+func (q *sender) sendPayloads(payloads []*payload) (done, retries uint64) {
+	var wg sync.WaitGroup
+	for _, p := range payloads {
+		q.climit <- struct{}{}
+		wg.Add(1)
+		go func(p *payload) {
+			defer func() { <-q.climit }()
+			defer wg.Done()
+			req, err := p.httpRequest(q.cfg.url)
+			if err != nil {
+				log.Errorf("http.Request: %s", err)
+				return
+			}
+			start := time.Now()
+			err = q.do(req)
+			stats := &eventData{
+				bytes:          p.body.Len(),
+				count:          1,
+				duration:       time.Since(start),
+				err:            err,
+				connectionFill: float64(len(q.climit)) / float64(cap(q.climit)),
+			}
+			switch err.(type) {
+			case *retriableError:
+				// request failed again, but can be retried
+				q.enqueueAgain(p)
+				atomic.AddUint64(&retries, 1)
+				q.recordEvent(eventTypeRetry, stats)
+				return
+			case nil:
+				// request was successful
+				atomic.AddUint64(&done, 1)
+				q.recordEvent(eventTypeSent, stats)
+			default:
+				// this is a fatal error, we have to drop this payload
+				q.recordEvent(eventTypeFailed, stats)
+			}
+			ppool.Put(p)
+		}(p)
+	}
+	wg.Wait()
+	return
+}
+
+// recordEvent records that event t has happened and attaches it the given data.
+func (q *sender) recordEvent(t eventType, data *eventData) {
+	if recorder := q.cfg.recorder; recorder != nil {
+		data.host = q.cfg.url.Hostname()
+		recorder.recordEvent(t, data)
+	}
+}
+
+// waitEmpty waits up to 5 seconds for the queue to reach an empty state and for all scheduling
+// to complete before returning.
+func (q *sender) waitEmpty() {
+	timeout := time.After(5 * time.Second)
+	ticker := time.NewTicker(100 * time.Millisecond)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-timeout:
+			return
+		case <-ticker.C:
+			q.mu.Lock()
+			drained := q.list.Len() == 0 && !q.scheduled
+			q.mu.Unlock()
+			if drained {
+				return
+			}
+		}
+	}
+}
+
+// enqueue enqueues the given payload. If there is no room in the queue, it drops oldest
+// payloads until there is.
+func (q *sender) enqueue(p *payload) {
+	q.mu.Lock()
+	defer q.mu.Unlock()
+	q.enqueueLocked(p)
+}
+
+// enqueueLocked adds p onto the queue, making room if needed.
+func (q *sender) enqueueLocked(p *payload) {
+	size := p.body.Len()
+	for q.size+size > maxQueueSize {
+		// make room
+		v := q.list.Remove(q.list.Front())
+		size := v.(*payload).body.Len()
+		q.size -= size
+		q.recordEvent(eventTypeDropped, &eventData{bytes: size, count: 1})
```

Comment:
> Since we already have the watchdog (which refuses payloads when memory gets too high), could we instead rely on that mechanism to drop payloads when we get too backed up here? 

I think we shouldn't always use "watchdog" as a generic motive for having unbounded things. We don't want anything to be unbounded in the application. Watchdog is an "embarrassing" system which should not be needed in well designed program and is something we plan to get rid of sooner or later. I don't think we should have unbounded queues, anywhere. Configurable, maybe, unbounded, no.

> From a user's perspective, the easiest way to think about agent configuration is to say "I will give the agent at most X memory to use".

I completely agree. That's why (as you may have noticed), this implementation has 100% less configuration than the previous one, which had a lot of knobs we've never used. My suggestion (as in a previous comment) was to use 12.8% of `apm_config.max_memory` for this queue (which is by default 64M). We can discuss if the number needs to be bigger, but in my tests, and given past history, it seemed fine. We can always reassess it later on.

I'm also happy to have a dedicated setting for this if we think it's a better idea.

Reply:
> I think we shouldn't always use "watchdog" as a generic motive for having unbounded things. We don't want anything to be unbounded in the application. Watchdog is an "embarrassing" system which should not be needed in a well designed program and is something we plan to get rid of sooner or later. I don't think we should have unbounded queues, anywhere. Configurable, maybe, unbounded, no.

Good point. I guess "unbounded" in this case is only acceptable assuming watchdog is what we rely on to give it a practical bound. But I agree, watchdog is kind of a hack, so maybe it's best not to rely on that as the primary way to enforce how big the "pending payloads" queue is. 

> I completely agree. That's why (as you may have noticed), this implementation has 100% less configuration than the previous one, which had a lot of knobs we've never used. My suggestion (as in a previous comment) was to use 12.8% of `apm_config.max_memory` for this queue (which is by default 64M). We can discuss if the number needs to be bigger, but in my tests, and given past history, it seemed fine. We can always reassess it later on.
> 
> I'm also happy to have a dedicated setting for this if we think it's a better idea.

Why not have it be effectively 100% of max_memory? If max_memory is being used as a proxy for "this is how much memory I want to devote to pending payloads", then it should be able to be bigger? 

Main concern is that for very high load agents that setting might only be a couple of seconds worth of payloads, so for brief network/service hiccups if it's too small we may end up dropping payloads that we could have saved had the queue been bigger.